x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java

```diff
@@ -0,0 +1,146 @@
+        this.appender = new DeprecationIndexingAppender("deprecation_indexing_appender", filter, ecsLayout, consumer);
+    }
+
+    @Override
+    protected void doStart() {
+        this.appender.start();
+        Loggers.addAppender(LogManager.getLogger("org.elasticsearch.deprecation"), this.appender);
+    }
+
+    @Override
+    protected void doStop() {
+        Loggers.removeAppender(LogManager.getLogger("org.elasticsearch.deprecation"), this.appender);
+        this.appender.stop();
+    }
+
+    @Override
+    protected void doClose() {
+        this.processor.close();
+    }
+
+    /**
+     * Listens for changes to the cluster state, in order to know whether to toggle indexing
+     * and to set the cluster UUID and node ID. These can't be set in the constructor because
+     * the initial cluster state won't be set yet.
+     *
+     * @param event the cluster state event to process
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        final boolean newEnabled = WRITE_DEPRECATION_LOGS_TO_INDEX.get(state.getMetadata().settings());
+        if (appender.isEnabled() != newEnabled) {
+            // We've flipped from disabled to enabled. Make sure we start with a clean cache of
+            // previously-seen keys, otherwise we won't index anything.
+            if (newEnabled) {
+                this.filter.reset();
+            }
+            appender.setEnabled(newEnabled);
+        }
+    }
+
+    /**
+     * Constructs a bulk processor for writing documents
+     *
+     * @param client   the client to use
+     * @param settings the settings to use
+     * @return an initialised bulk processor
+     */
+    private BulkProcessor getBulkProcessor(Client client, Settings settings) {
+        final OriginSettingClient originSettingClient = new OriginSettingClient(client, ClientHelper.DEPRECATION_ORIGIN);
+        final BulkProcessor.Listener listener = new DeprecationBulkListener();
+
+        // This configuration disables the size count and size thresholds,
+        // and instead uses a scheduled flush only. This means that calling
+        // processor.add() will not block the calling thread.
+        return BulkProcessor.builder(originSettingClient::bulk, listener)
+            .setBackoffPolicy(BackoffPolicy.exponentialBackoff(TimeValue.timeValueMillis(1000), 3))
+            .setConcurrentRequests(Math.max(2, EsExecutors.allocatedProcessors(settings)))
+            .setBulkActions(-1)
+            .setBulkSize(new ByteSizeValue(-1, ByteSizeUnit.BYTES))
+            .setFlushInterval(TimeValue.timeValueSeconds(5))
+            .build();
+    }
+
+    private static class DeprecationBulkListener implements BulkProcessor.Listener {
+        @Override
+        public void beforeBulk(long executionId, BulkRequest request) {}
+
+        @Override
+        public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {}
```

Comment:
Can you add this to the body of the afterBulk() method to introspect if there are bulk item failures?

```
long items = request.numberOfActions();
if (logger.isTraceEnabled()) {
   logger.trace("indexed [{}] deprecation documents into [{}]", items,
   Arrays.stream(response.getItems())
      .map(BulkItemResponse::getIndex)
      .distinct()
      .collect(Collectors.joining(",")));
}

if (response.hasFailures()) {
   Map<String, String> failures = Arrays.stream(response.getItems())
      .filter(BulkItemResponse::isFailed)
      .collect(Collectors.toMap(BulkItemResponse::getId, BulkItemResponse::getFailureMessage));
   logger.error("failures: [{}]", failures);
}
```